File: parser.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdexcept>

enum class TokenType {
    Keyword,
    Identifier,
    Number,
    Symbol,
    EndOfFile,
    Unknown
};

struct Token {
    TokenType type;
    std::string value;

    Token(TokenType t, std::string v) : type(t), value(std::move(v)) {}
};

class Lexer {
public:
    Lexer(const std::string &source) : source(source), position(0) {}

    std::vector<Token> tokenize() {
        std::vector<Token> tokens;
        while (position < source.length()) {
            char current = source[position];
            if (isspace(current)) {
                position++;
            } else if (isalpha(current)) {
                tokens.push_back(tokenizeIdentifierOrKeyword());
            } else if (isdigit(current)) {
                tokens.push_back(tokenizeNumber());
            } else {
                tokens.push_back(tokenizeSymbol());
            }
        }
        tokens.emplace_back(TokenType::EndOfFile, "");
        return tokens;
    }

private:
    const std::string source;
    size_t position;

    Token tokenizeIdentifierOrKeyword() {
        size_t start = position;
        while (position < source.length() && (isalnum(source[position]) || source[position] == '_')) {
            position++;
        }
        std::string value = source.substr(start, position - start);
        if (value == "if" || value == "else" || value == "while") {  // Add more keywords here
            return Token(TokenType::Keyword, value);
        }
        return Token(TokenType::Identifier, value);
    }

    Token tokenizeNumber() {
        size_t start = position;
        while (position < source.length() && isdigit(source[position])) {
            position++;
        }
        std::string value = source.substr(start, position - start);
        return Token(TokenType::Number, value);
    }

    Token tokenizeSymbol() {
        char current = source[position];
        position++;
        return Token(TokenType::Symbol, std::string(1, current));
    }
};

class ASTNode {};
class ASTExpression : public ASTNode {
public:
    Token token;

    ASTExpression(const Token &token) : token(token) {}
};

class ASTStatement : public ASTNode {};
class ASTIfStatement : public ASTStatement {
public:
    ASTExpression condition;
    ASTStatement *body;

    ASTIfStatement(const ASTExpression &condition, ASTStatement *body) : condition(condition), body(body) {}
};

class ASTWhileStatement : public ASTStatement {
public:
    ASTExpression condition;
    ASTStatement *body;

    ASTWhileStatement(const ASTExpression &condition, ASTStatement *body) : condition(condition), body(body) {}
};

class ASTAssignment : public ASTStatement {
public:
    Token variable;
    ASTExpression expression;

    ASTAssignment(const Token &variable, const ASTExpression &expression) : variable(variable), expression(expression) {}
};

class Parser {
public:
    Parser(const std::vector<Token> &tokens) : tokens(tokens), position(0) {}

    ASTStatement *parse() {
        return parseStatement();
    }

private:
    const std::vector<Token> tokens;
    size_t position;

    ASTStatement *parseStatement() {
        Token token = tokens[position];
        if (token.type == TokenType::Keyword) {
            if (token.value == "if") return parseIfStatement();
            if (token.value == "while") return parseWhileStatement();
        } else if (token.type == TokenType::Identifier) {
            return parseAssignment();
        }
        throw std::runtime_error("Unexpected token: " + token.value);
    }

    ASTIfStatement *parseIfStatement() {
        position++; // consume 'if'
        ASTExpression condition = parseExpression();
        ASTStatement *body = parseStatement();
        return new ASTIfStatement(condition, body);
    }

    ASTWhileStatement *parseWhileStatement() {
        position++; // consume 'while'
        ASTExpression condition = parseExpression();
        ASTStatement *body = parseStatement();
        return new ASTWhileStatement(condition, body);
    }

    ASTAssignment *parseAssignment() {
        Token variable = tokens[position];
        position++; // consume variable
        Token equals = tokens[position];
        if (equals.value != "=") throw std::runtime_error("Expected '='");
        position++; // consume '='
        ASTExpression expression = parseExpression();
        return new ASTAssignment(variable, expression);
    }

    ASTExpression parseExpression() {
        Token token = tokens[position];
        position++;
        return ASTExpression(token);
    }
};

int main() {
    std::string code = "if x == 10 while y != 20";
    Lexer lexer(code);
    std::vector<Token> tokens = lexer.tokenize();

    Parser parser(tokens);
    ASTStatement *ast = parser.parse();

    std::cout << "Parsing completed." << std::endl;

    return 0;
}

File: parser.cs
using System;
using System.Collections.Generic;

public enum TokenType {
    Keyword,
    Identifier,
    Number,
    Symbol,
    EndOfFile,
    Unknown
}

public class Token {
    public TokenType Type { get; }
    public string Value { get; }

    public Token(TokenType type, string value) {
        Type = type;
        Value = value;
    }
}

public class Lexer {
    private readonly string _source;
    private int _position;

    public Lexer(string source) {
        _source = source;
        _position = 0;
    }

    public List<Token> Tokenize() {
        var tokens = new List<Token>();
        while (_position < _source.Length) {
            var current = _source[_position];
            if (char.IsWhiteSpace(current)) {
                _position++;
            } else if (char.IsLetter(current)) {
                tokens.Add(TokenizeIdentifierOrKeyword());
            } else if (char.IsDigit(current)) {
                tokens.Add(TokenizeNumber());
            } else {
                tokens.Add(TokenizeSymbol());
            }
        }
        tokens.Add(new Token(TokenType.EndOfFile, string.Empty));
        return tokens;
    }

    private Token TokenizeIdentifierOrKeyword() {
        var start = _position;
        while (_position < _source.Length && (char.IsLetterOrDigit(_source[_position]) || _source[_position] == '_')) {
            _position++;
        }
        var value = _source.Substring(start, _position - start);
        return value switch {
            "if" or "else" or "while" => new Token(TokenType.Keyword, value),
            _ => new Token(TokenType.Identifier, value)
        };
    }

    private Token TokenizeNumber() {
        var start = _position;
        while (_position < _source.Length && char.IsDigit(_source[_position])) {
            _position++;
        }
        var value = _source.Substring(start, _position - start);
        return new Token(TokenType.Number, value);
    }

    private Token TokenizeSymbol() {
        var value = _source[_position].ToString();
        _position++;
        return new Token(TokenType.Symbol, value);
    }
}

public abstract class ASTNode {}

public class ASTExpression : ASTNode {
    public Token Token { get; }

    public ASTExpression(Token token) {
        Token = token;
    }
}

public class ASTStatement : ASTNode {}

public class ASTIfStatement : ASTStatement {
    public ASTExpression Condition { get; }
    public ASTStatement Body { get; }

    public ASTIfStatement(ASTExpression condition, ASTStatement body) {
        Condition = condition;
        Body = body;
    }
}

public class ASTWhileStatement : ASTStatement {
    public ASTExpression Condition { get; }
    public ASTStatement Body { get; }

    public ASTWhileStatement(ASTExpression condition, ASTStatement body) {
        Condition = condition;
        Body = body;
    }
}

public class ASTAssignment : ASTStatement {
    public Token Variable { get; }
    public ASTExpression Expression { get; }

    public ASTAssignment(Token variable, ASTExpression expression) {
        Variable = variable;
        Expression = expression;
    }
}

public class Parser {
    private readonly List<Token> _tokens;
    private int _position;

    public Parser(List<Token> tokens) {
        _tokens = tokens;
        _position = 0;
    }

    public ASTStatement Parse() {
        return ParseStatement();
    }

    private ASTStatement ParseStatement() {
        var token = _tokens[_position];
        if (token.Type == TokenType.Keyword) {
            if (token.Value == "if") return ParseIfStatement();
            if (token.Value == "while") return ParseWhileStatement();
        } else if (token.Type == TokenType.Identifier) {
            return ParseAssignment();
        }
        throw new Exception($"Unexpected token: {token.Value}");
    }

    private ASTIfStatement ParseIfStatement() {
        _position++; // consume 'if'
        var condition = ParseExpression();
        var body = ParseStatement();
        return new ASTIfStatement(condition, body);
    }

    private ASTWhileStatement ParseWhileStatement() {
        _position++; // consume 'while'
        var condition = ParseExpression();
        var body = ParseStatement();
        return new ASTWhileStatement(condition, body);
    }

    private ASTAssignment ParseAssignment() {
        var variable = _tokens[_position];
        _position++; // consume variable
        var equals = _tokens[_position];
        if (equals.Value != "=") throw new Exception("Expected '='");
        _position++; // consume '='
        var expression = ParseExpression();
        return new ASTAssignment(variable, expression);
    }

    private ASTExpression ParseExpression() {
        var token = _tokens[_position];
        _position++;
        return new ASTExpression(token);
    }
}

public class SharkaraInterpreter {
    public static void Main(string[] args) {
        string code = "if x == 10 while y != 20";
        Lexer lexer = new Lexer(code);
        List<Token> tokens = lexer.Tokenize();

        Parser parser = new Parser(tokens);
        ASTStatement ast = parser.Parse();

        Console.WriteLine("Parsing completed.");
    }
}
